xdsl/dialects/memref.py

```diff
@@ -51,9 +51,7 @@ def from_params(
         return MemRefType([shape, referenced_type])
 
 
-_UnrankedMemrefTypeElems = TypeVar("_UnrankedMemrefTypeElems",
-                                   bound=Attribute,
-                                   covariant=True)
+_UnrankedMemrefTypeElems = TypeVar("_UnrankedMemrefTypeElems", bound=Attribute)
```

Comment:
I don't get any more type errors when I remove the `covariant=True`, so I guess this is ok? Functionality isn't affected

Reply:
We should try to make this work with `covariant=True`. What are the errors we get if we keep it?
The reason we want to make this covariant is that when we have a function expecting a `UnrankedMemrefType[A]` and we give a `UnrankedMemrefType[B]`, with `B` a subclass of `A`.

Reply:
Not compatible with its use below. `List` argument cannot be covariant, since it's mutable

Reply:
<img width="1005" alt="Screenshot 2023-02-18 at 18 54 31" src="https://user-images.githubusercontent.com/910320/219883256-620435df-ff8f-4fd1-bb13-59d3f39d88e0.png">

Reply:
So, fun fact, I think we need to use 2 type variables.
One for the class definition, and one for this method, where the one for this method is invariant (the default option).

Reply:
Ah I can try that
- [ ] try that

Reply:
Let's try that in another PR

Reply:
Actually added it in here